lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,178 @@
+# (C) British Crown Copyright 2017, Met Office
+#
+# This file is part of Iris.
+#
+# Iris is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Iris is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Routines for putting data on new strata (aka. isosurfaces), often in the
+Z direction.
+
+"""
+import copy
+from functools import partial
+
+import numpy as np
+import stratify
+
+from iris.cube import Cube
+from iris.coords import Coord, AuxCoord, DimCoord
+
+
+def _copy_coords_without_dim(cube, result, z_dim):
+    """
+    Helper function to copy across non-z coordinates between cubes.
+
+    Parameters
+    ----------
+    cube : iris.cube.Cube instance
+        Incoming cube containing the coordinates to be copied.
+    result : iris.cube.Cube instance
+        Incoming cube for the coordinates to be copied to.
+    z_dim : Dimension within the cube of the z-dimension
+        The dimension will not be copied, as the incoming z-dimension
+        is model level_height, whilst the outgoing z-dimension is pressure.
+
+    Returns
+    -------
+    Cube
+        A single cube with the non-z dimensions added to
+        incoming cube.
+
+    """
+    # Copy across non-Z coordinates.
+    for coord in cube.dim_coords:
+        dim, = cube.coord_dims(coord)
+        if dim != z_dim:
+            result.add_dim_coord(coord.copy(), dim)
+    for coord in cube.aux_coords:
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+    for coord in cube.derived_coords:
+        dims = cube.coord_dims(coord)
+        if z_dim not in dims:
+            result.add_aux_coord(coord.copy(), dims)
+
+
+def relevel(phenom, level_data, target_levels, interp_direction=0,
+            interpolator=None):
+    """
+    Interpolate the given phenomenon onto the given levels of level_data.
+
+    For example, suppose we have two datasets `P(i,j,k)` and `H(i,j,k)`
+    and we want ``P(i,j,H)`` we call relevel with
+    ``phenom=P`` and ``level_data=H`` and ``target_levels`` being
+    an array of the values of ``H`` we would like.
+
+    This routine is especially useful for computing isosurfaces of
+    phenomenon that are generally monotonic in the direction of interpolation
+    (such as height/pressure and salinity/depth).
+
+    Parameters
+    ----------
+    phenom
+        The data to be re-levelled.
+    
+    level_data : iris.cube.Cube or iris.coord.Coord
```

Comment:
level_data might be better named source_levels you think?? (I guess data signifies that it might also be passed as a cube??)